File: src/main/java/com/dh/ondot/member/domain/Member.java
package com.dh.ondot.member.domain;

import com.dh.ondot.core.AggregateRoot;
import com.dh.ondot.core.domain.BaseTimeEntity;
import com.dh.ondot.core.util.DateTimeUtils;
import com.dh.ondot.member.domain.enums.MapProvider;
import com.dh.ondot.member.domain.enums.OauthProvider;
import com.dh.ondot.schedule.domain.enums.AlarmMode;
import com.dh.ondot.schedule.domain.vo.Snooze;
import com.dh.ondot.schedule.domain.vo.Sound;
import jakarta.persistence.*;
import lombok.*;

import java.time.Instant;

@AggregateRoot
@Entity
@Getter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Table(name = "members")
public class Member extends BaseTimeEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "member_id")
    private Long id;

    @Column(name = "nickname", length = 50)
    private String nickname;

    @Column(name = "email", unique = true, nullable = false)
    private String email;

    @Embedded
    private OauthInfo oauthInfo;

    @Column(name = "latest_preparation_alarm_id")
    private Long latestPreparationAlarmId;

    @Column(name = "latest_departure_alarm_id")
    private Long latestDepartureAlarmId;

    @Column(name = "preparation_time")
    private Integer preparationTime;

    @Enumerated(EnumType.STRING)
    @Column(name = "default_alarm_mode")
    private AlarmMode defaultAlarmMode;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "isSnoozeEnabled", column = @Column(name = "default_is_snooze_enabled")),
            @AttributeOverride(name = "snoozeInterval", column = @Column(name = "default_snooze_interval")),
            @AttributeOverride(name = "snoozeCount", column = @Column(name = "default_snooze_count"))
    })
    private Snooze snooze;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "soundCategory", column = @Column(name = "default_sound_category")),
            @AttributeOverride(name = "ringTone", column = @Column(name = "default_ring_tone")),
            @AttributeOverride(name = "volume", column = @Column(name = "default_volume"))
    })
    private Sound sound;

    @Enumerated(EnumType.STRING)
    @Column(name = "map_provider")
    private MapProvider mapProvider;

    @Column(name = "deleted_at")
    private Instant deletedAt;

    public static Member registerWithOauth(String email, OauthProvider oauthProvider, String oauthProviderId) {
        return Member.builder()
                .email(email)
                .oauthInfo(OauthInfo.of(oauthProvider, oauthProviderId))
                .build();
    }

    public void updateOnboarding(
            Integer preparationTime, String alarmMode,
            boolean isSnoozeEnabled, Integer snoozeInterval, Integer snoozeCount,
            String soundCategory, String ringTone, Double volume
    ) {
        this.preparationTime = preparationTime;
        this.defaultAlarmMode = AlarmMode.from(alarmMode);
        this.snooze = Snooze.of(isSnoozeEnabled, snoozeInterval, snoozeCount);
        this.sound = Sound.of(soundCategory, ringTone, volume);
    }

    public void updateMapProvider(String mapProvider) {
        this.mapProvider = MapProvider.from(mapProvider);
    }

    public boolean checkOnboardingCompleted() {
        return preparationTime != null;
    }

    public void deactivate() {
        this.email = "deactivated-" + this.id + "@ondot.com";
        this.oauthInfo = OauthInfo.of(this.getOauthInfo().getOauthProvider(), "0");
        this.deletedAt = DateTimeUtils.nowSeoulInstant();
    }
}

File: src/main/kotlin/com/dh/ondot/member/domain/Member.kt
package com.dh.ondot.member.domain

import com.dh.ondot.core.AggregateRoot
import com.dh.ondot.core.BaseTimeEntity
import com.dh.ondot.member.domain.enums.MapProvider
import com.dh.ondot.member.domain.enums.OauthProvider
import com.dh.ondot.schedule.domain.enums.AlarmMode
import com.dh.ondot.schedule.domain.vo.Snooze
import com.dh.ondot.schedule.domain.vo.Sound
import jakarta.persistence.*
import java.time.Instant

@AggregateRoot
@Entity
@Table(
    name = "members",
    uniqueConstraints = [
        UniqueConstraint(
            name = "uk_member_oauth",
            columnNames = ["oauth_provider_id", "oauth_provider"]
        )
    ]
)
class Member(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "member_id")
    val id: Long = 0L,

    @Column(name = "nickname", length = 50)
    var nickname: String? = null,

    @Column(name = "email", nullable = false)
    val email: String,

    @Embedded
    var oauthInfo: OauthInfo,

    @Column(name = "latest_preparation_alarm_id")
    var latestPreparationAlarmId: Long? = null,

    @Column(name = "latest_departure_alarm_id")
    var latestDepartureAlarmId: Long? = null,

    @Column(name = "preparation_time")
    var preparationTime: Int? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "default_alarm_mode")
    var defaultAlarmMode: AlarmMode? = null,

    @Embedded
    @AttributeOverrides(
        AttributeOverride(name = "isSnoozeEnabled", column = Column(name = "default_is_snooze_enabled")),
        AttributeOverride(name = "snoozeInterval", column = Column(name = "default_snooze_interval")),
        AttributeOverride(name = "snoozeCount", column = Column(name = "default_snooze_count"))
    )
    var snooze: Snooze? = null,

    @Embedded
    @AttributeOverrides(
        AttributeOverride(name = "soundCategory", column = Column(name = "default_sound_category")),
        AttributeOverride(name = "ringTone", column = Column(name = "default_ring_tone")),
        AttributeOverride(name = "volume", column = Column(name = "default_volume"))
    )
    var sound: Sound? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "map_provider")
    var mapProvider: MapProvider? = null,

    @Column(name = "deleted_at")
    var deletedAt: Instant? = null,

    @Column(name = "daily_reminder_enabled", nullable = false, columnDefinition = "TINYINT(1)")
    var dailyReminderEnabled: Boolean = true,
) : BaseTimeEntity() {

    fun updateOnboarding(
        preparationTime: Int,
        alarmMode: String,
        isSnoozeEnabled: Boolean,
        snoozeInterval: Int?,
        snoozeCount: Int?,
        soundCategory: String,
        ringTone: String,
        volume: Double?,
    ) {
        this.preparationTime = preparationTime
        this.defaultAlarmMode = AlarmMode.from(alarmMode)
        this.snooze = Snooze.of(isSnoozeEnabled, snoozeInterval, snoozeCount)
        this.sound = Sound.of(soundCategory, ringTone, volume)
    }

    fun updateMapProvider(mapProvider: String) {
        this.mapProvider = MapProvider.from(mapProvider)
    }

    fun updatePreparationTime(preparationTime: Int?) {
        this.preparationTime = preparationTime
    }

    fun isNewMember(): Boolean = preparationTime == null

    fun updateDailyReminderEnabled(enabled: Boolean) {
        this.dailyReminderEnabled = enabled
    }

    companion object {
        @JvmStatic
        fun registerWithOauth(email: String, oauthProvider: OauthProvider, oauthProviderId: String): Member =
            Member(
                email = email,
                oauthInfo = OauthInfo.of(oauthProvider, oauthProviderId),
            )
    }
}
